File: messaging/signal/calling/media/srtp_kdf.c
/*
 * srtp_kdf.c - RingRTC 1:1 Signal-call SRTP key derivation, OpenSSL/C port for the teamsm-style
 * Signal call mediator. Mirrors ringrtc connection.rs::negotiate_srtp_keys exactly; verified byte-
 * for-byte against messaging/signal/calling/media/srtp_kdf.py (RFC7748 X25519 + RFC5869 HKDF).
 *
 *   shared = X25519(local_priv, remote_pub)                 // reject all-zero (non-contributory)
 *   okm    = HKDF-SHA256(ikm=shared, salt=32*0x00,
 *                        info="Signal_Calling_20200807_SignallingDH_SRTPKey_KDF"+caller_id+callee_id, L=88)
 *   offer_key=okm[0:32] offer_salt=okm[32:44] answer_key=okm[44:76] answer_salt=okm[76:88]  // AEAD_AES_256_GCM
 *
 * OpenSSL 1.1.1+ (device has 1.1.1w). Build/test on host:
 *   cc srtp_kdf.c -o srtp_kdf -lcrypto && ./srtp_kdf   # runs the same vector as srtp_kdf.py --demo
 */
#include <stdio.h>
#include <string.h>
#include <openssl/evp.h>
#include <openssl/kdf.h>
#include <openssl/err.h>

#define KEY_SIZE 32
#define SALT_SIZE 12
#define OKM_LEN (KEY_SIZE + SALT_SIZE + KEY_SIZE + SALT_SIZE)   /* 88 */
static const char *HKDF_INFO_PREFIX = "Signal_Calling_20200807_SignallingDH_SRTPKey_KDF";

typedef struct {
    unsigned char offer_key[KEY_SIZE],  offer_salt[SALT_SIZE];   /* caller (offerer) sends */
    unsigned char answer_key[KEY_SIZE], answer_salt[SALT_SIZE];  /* callee (answerer) sends */
} signal_srtp_keys;

/* X25519 shared secret. Returns 0 on success. */
static int x25519_shared(const unsigned char local_priv[32], const unsigned char remote_pub[32],
                         unsigned char out_shared[32]) {
    int rc = -1;
    EVP_PKEY *loc = EVP_PKEY_new_raw_private_key(EVP_PKEY_X25519, NULL, local_priv, 32);
    EVP_PKEY *rem = EVP_PKEY_new_raw_public_key(EVP_PKEY_X25519, NULL, remote_pub, 32);
    EVP_PKEY_CTX *ctx = loc ? EVP_PKEY_CTX_new(loc, NULL) : NULL;
    size_t len = 32;
    if (loc && rem && ctx && EVP_PKEY_derive_init(ctx) == 1 &&
        EVP_PKEY_derive_set_peer(ctx, rem) == 1 &&
        EVP_PKEY_derive(ctx, out_shared, &len) == 1 && len == 32)
        rc = 0;
    EVP_PKEY_CTX_free(ctx); EVP_PKEY_free(loc); EVP_PKEY_free(rem);
    return rc;
}

/* Derive the local X25519 public key from a private key (for the Answer's public_key). 0 on success. */
int signal_x25519_public_from_private(const unsigned char local_priv[32], unsigned char out_pub[32]) {
    size_t len = 32;
    EVP_PKEY *loc = EVP_PKEY_new_raw_private_key(EVP_PKEY_X25519, NULL, local_priv, 32);
    int rc = (loc && EVP_PKEY_get_raw_public_key(loc, out_pub, &len) == 1 && len == 32) ? 0 : -1;
    EVP_PKEY_free(loc);
    return rc;
}

/* The full derivation. caller/callee identity keys are the two parties' Signal identity PUBLIC keys
 * (libsignal 33-byte 0x05-prefixed form). Returns 0 on success, -1 on error (incl. non-contributory). */
int signal_negotiate_srtp_keys(const unsigned char local_priv[32], const unsigned char remote_pub[32],
                               const unsigned char *caller_id, size_t caller_id_len,
                               const unsigned char *callee_id, size_t callee_id_len,
                               signal_srtp_keys *out) {
    unsigned char shared[32], okm[OKM_LEN], zero[32] = {0}, salt[32] = {0};
    if (x25519_shared(local_priv, remote_pub, shared) != 0) return -1;
    if (memcmp(shared, zero, 32) == 0) return -1;   /* non-contributory */

    size_t info_len = strlen(HKDF_INFO_PREFIX) + caller_id_len + callee_id_len;
    unsigned char *info = OPENSSL_malloc(info_len);
    if (!info) return -1;
    size_t off = strlen(HKDF_INFO_PREFIX);
    memcpy(info, HKDF_INFO_PREFIX, off);
    memcpy(info + off, caller_id, caller_id_len); off += caller_id_len;
    memcpy(info + off, callee_id, callee_id_len);

    int rc = -1;
    EVP_PKEY_CTX *h = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);
    size_t olen = OKM_LEN;
    if (h && EVP_PKEY_derive_init(h) == 1 &&
        EVP_PKEY_CTX_set_hkdf_md(h, EVP_sha256()) == 1 &&
        EVP_PKEY_CTX_set1_hkdf_salt(h, salt, 32) == 1 &&
        EVP_PKEY_CTX_set1_hkdf_key(h, shared, 32) == 1 &&
        EVP_PKEY_CTX_add1_hkdf_info(h, info, info_len) == 1 &&
        EVP_PKEY_derive(h, okm, &olen) == 1 && olen == OKM_LEN) {
        memcpy(out->offer_key,   okm,                              KEY_SIZE);
        memcpy(out->offer_salt,  okm + KEY_SIZE,                   SALT_SIZE);
        memcpy(out->answer_key,  okm + KEY_SIZE + SALT_SIZE,       KEY_SIZE);
        memcpy(out->answer_salt, okm + KEY_SIZE + SALT_SIZE + KEY_SIZE, SALT_SIZE);
        rc = 0;
    }
    EVP_PKEY_CTX_free(h);
    OPENSSL_clear_free(info, info_len);
    OPENSSL_cleanse(shared, 32); OPENSSL_cleanse(okm, OKM_LEN);
    return rc;
}

#ifdef SRTP_KDF_TEST
static void hx(const char *l, const unsigned char *b, size_t n) {
    printf("%s", l); for (size_t i = 0; i < n; i++) printf("%02x", b[i]); printf("\n");
}
static int fromhex(const char *h, unsigned char *out, size_t n) {
    for (size_t i = 0; i < n; i++) if (sscanf(h + 2*i, "%2hhx", &out[i]) != 1) return -1;
    return 0;
}
int main(void) {
    /* Same fixed vector as srtp_kdf.py's demo — must match its EXPECT line byte-for-byte. */
    unsigned char priv[32], alan[32], cid[33], eid[33];
    fromhex("0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20", priv, 32);
    fromhex("364fb3e8d71d17b656aca498ca2c4e9ffacfab68b6a7255507a8ca1d58637361", alan, 32);
    fromhex("05aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", cid, 33);
    fromhex("05bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb", eid, 33);
    unsigned char pub[32];
    signal_x25519_public_from_private(priv, pub);
    hx("our_public  = ", pub, 32);
    signal_srtp_keys k;
    if (signal_negotiate_srtp_keys(priv, alan, cid, 33, eid, 33, &k) != 0) {
        fprintf(stderr, "derivation FAILED\n"); return 1;
    }
    hx("offer_key   = ", k.offer_key, 32);
    hx("offer_salt  = ", k.offer_salt, 12);
    hx("answer_key  = ", k.answer_key, 32);
    hx("answer_salt = ", k.answer_salt, 12);
    return 0;
}
#endif

File: messaging/signal/calling/media/srtp_kdf.py
#!/usr/bin/env python3
"""
srtp_kdf.py - reference implementation of RingRTC's 1:1 Signal-call SRTP key derivation
(ringrtc src/rust/src/core/connection.rs::negotiate_srtp_keys), for verifying the C port.

Signal 1:1 call media = AEAD_AES_256_GCM SRTP whose keys come from an X25519 DH on the
public_key carried in the offer/answer opaque (ConnectionParametersV4). NOT DTLS.

Derivation (verbatim from RingRTC):
  shared = X25519(local_secret, remote_public_key)          # reject non-contributory (all-zero)
  okm    = HKDF-SHA256(ikm=shared, salt=32*0x00,
                       info="Signal_Calling_20200807_SignallingDH_SRTPKey_KDF"
                            + caller_identity_key + callee_identity_key,
                       L=32+12+32+12 = 88)
  offer_key  = okm[0:32]   offer_salt  = okm[32:44]   # OFFERER (caller) sends with this
  answer_key = okm[44:76]  answer_salt = okm[76:88]   # ANSWERER (callee) sends with this

For an INCOMING call we are the callee/answerer:
  - DECRYPT the caller's audio with (offer_key, offer_salt)
  - ENCRYPT our audio with (answer_key, answer_salt)
Both peers compute the SAME okm (info is always caller_id then callee_id, role-independent).

identity keys: the two parties' Signal identity PUBLIC keys. libsignal serialises an IdentityKey
as 33 bytes = 0x05 (DJB type) || 32-byte X25519 pub. UNVERIFIED whether RingRTC passes the 33-byte
or 32-byte form here - flagged; confirm on a live call. Toggle IDKEY_33 below to test both.
"""
import sys, binascii
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes

HKDF_INFO_PREFIX = b"Signal_Calling_20200807_SignallingDH_SRTPKey_KDF"
KEY_SIZE, SALT_SIZE = 32, 12          # AEAD_AES_256_GCM

def negotiate_srtp_keys(local_priv32: bytes, remote_pub32: bytes,
                        caller_id: bytes, callee_id: bytes):
    if len(local_priv32) != 32 or len(remote_pub32) != 32:
        raise ValueError("local priv and remote pub must be 32 bytes")
    sk = X25519PrivateKey.from_private_bytes(local_priv32)
    pk = X25519PublicKey.from_public_bytes(remote_pub32)
    shared = sk.exchange(pk)                       # X25519, RFC 7748
    if shared == b"\x00" * 32:                     # non-contributory -> RingRTC rejects
        raise ValueError("non-contributory shared secret (low-order remote key)")
    info = HKDF_INFO_PREFIX + caller_id + callee_id
    okm = HKDF(algorithm=hashes.SHA256(), length=KEY_SIZE + SALT_SIZE + KEY_SIZE + SALT_SIZE,
               salt=b"\x00" * 32, info=info).derive(shared)
    return {
        "offer_key":  okm[0:32],   "offer_salt":  okm[32:44],   # caller sends
        "answer_key": okm[44:76],  "answer_salt": okm[76:88],   # callee sends
    }

def local_pub_from_priv(local_priv32: bytes) -> bytes:
    from cryptography.hazmat.primitives import serialization
    return X25519PrivateKey.from_private_bytes(local_priv32).public_key().public_bytes(
        serialization.Encoding.Raw, serialization.PublicFormat.Raw)

# ---------------------------------------------------------------- self-tests (RFC vectors) -----
def _selftest():
    h = binascii.unhexlify
    # RFC 7748 s6.1 X25519 DH known-answer
    a_priv = h("77076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c2a")
    b_pub  = h("de9edb7d7b7dc1b4d35b61c2ece435373f8343c85b78674dadfc7e146f882b4f")
    want   = h("4a5d9d5ba4ce2de1728e3bf480350f25e07e21c947d19e3376f09b3c1e161742")
    got = X25519PrivateKey.from_private_bytes(a_priv).exchange(X25519PublicKey.from_public_bytes(b_pub))
    assert got == want, "X25519 RFC7748 vector FAILED"
    # RFC 5869 Test Case 1 HKDF-SHA256
    okm = HKDF(algorithm=hashes.SHA256(), length=42, salt=h("000102030405060708090a0b0c"),
               info=h("f0f1f2f3f4f5f6f7f8f9")).derive(h("0b"*22))
    assert okm == h("3cb25f25faacd57a90434f64d0362f2a2d2d0a90cf1a5a4c5db02d56ecc4c5bf34007208d5b887185865"), \
        "HKDF RFC5869 vector FAILED"
    print("[selftest] X25519 (RFC7748) + HKDF-SHA256 (RFC5869) vectors PASS -> primitives correct")

def _hx(b): return binascii.hexlify(b).decode()

if __name__ == "__main__":
    _selftest()
    # Demo on Alan's REAL captured offer public_key (2026-07-21 sigoffer.log). Local priv + identity
    # keys are PLACEHOLDERS (the real ones come from presage at call time); this only shows the
    # derivation runs deterministically end-to-end on real remote data.
    alan_pub = binascii.unhexlify("364fb3e8d71d17b656aca498ca2c4e9ffacfab68b6a7255507a8ca1d58637361")
    local_priv = bytes(range(1, 33))                    # placeholder ephemeral secret
    caller_id  = b"\x05" + b"\xAA"*32                    # placeholder Alan identity (33B, 0x05-prefixed)
    callee_id  = b"\x05" + b"\xBB"*32                    # placeholder our identity
    print("\n[demo] our ephemeral public (goes in Answer.ConnectionParametersV4.public_key):")
    print("   ", _hx(local_pub_from_priv(local_priv)))
    k = negotiate_srtp_keys(local_priv, alan_pub, caller_id, callee_id)
    print("[demo] derived SRTP keys (AEAD_AES_256_GCM) from Alan's public_key:")
    for name in ("offer_key", "offer_salt", "answer_key", "answer_salt"):
        print(f"    {name:11} = {_hx(k[name])}")
    print("  -> incoming call: DECRYPT caller audio with offer_key/salt; ENCRYPT ours with answer_key/salt")
    # emit a machine-readable line the C port can diff against
    print("\nVECTOR local_priv=%s remote_pub=%s caller_id=%s callee_id=%s" % (
        _hx(local_priv), _hx(alan_pub), _hx(caller_id), _hx(callee_id)))
    print("EXPECT offer_key=%s offer_salt=%s answer_key=%s answer_salt=%s" % (
        _hx(k["offer_key"]), _hx(k["offer_salt"]), _hx(k["answer_key"]), _hx(k["answer_salt"])))
